### ISSUE TITLE
Custom Rule Errors with `undefined`

### ISSUE THREAD

**Issue by nategoethel** [OnkarRuikar/markdownlint-rule-search-replace#56]
Here is our `.markdownlint.json` file with a custom rule, per the documentation:
<img width="538" alt="Screenshot 2023-02-09 at 12 29 39 PM" src="https://user-images.githubusercontent.com/44580393/217904887-2138eef9-0554-4c48-aecd-0e3744a4bc97.png">

This is erroring out when actually linting a file:
<img width="850" alt="Screenshot 2023-02-09 at 12 31 15 PM" src="https://user-images.githubusercontent.com/44580393/217905278-f0f3fdb2-edc6-4c36-bd89-7d1b41d6cad1.png">

I believe the error is occurring here, but I'm not able to debug it:

https://github.com/OnkarRuikar/markdownlint-rule-search-replace/blob/4ef9041f12e1309ca85693152683bdeced392be5/rule.js#L17-L21

**Comment by OnkarRuikar** [comment]
@nategoethel there are couple of issues with the configuration here:
- the option `searchPattern` needs to be in camel case.
- the regex doesn't start with `/` like : `/my-regex/gi`
- escape `\` with `\`: `/abc \\s def/gi`

Let me know if following code solves the issue.
```json
{
  "name": "bold angle brackets",
  "message": "Do not bold angle brackets",
  "searchPattern": "/(\\*\\*)([A-Z]+)(\\s)(\\>)(\\s)([A-Z]+)(\\*\\*)/gi",
  "replace": "**$2** > **$6**",
  "skipCode": true
}
```

![bug](https://user-images.githubusercontent.com/87750369/217981300-29b693d3-5a50-4cf0-828f-f4a654842934.png)

**Comment by nategoethel** [comment]
@OnkarRuikar thanks for the help. Can you clarify whether we need to point the `markdownlint` VSCode extension to this package? If so, how?

I updated the rule per your recommendation (in our `.markdownlint.json` file; I see you're using a `.markdownlint-cli2.jsonc` file), but I'm not seeing the rule be picked up.

**Comment by OnkarRuikar** [comment]
> @OnkarRuikar thanks for the help. Can you clarify whether we need to point the `markdownlint` VSCode extension to this package? If so, how?

Custom rules are not bundled with the extension. You have to download them separately.
Download the `rule.js` file from here and put it at suitable location in your project directory. Make sure you commit the file in project repo so that others can use it. Preferably use `.vscode/rule.js` location. Then update the config file using [these instructions](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint#markdownlint.customrules).

> I updated the rule per your recommendation (in our `.markdownlint.json` file; I see you're using a `.markdownlint-cli2.jsonc` file), but I'm not seeing the rule be picked up.

I tested this using [a NodeJS project](https://github.com/mdn/content/blob/6d2bbd133371731c0896cb62803db8251c81b864/package.json#L32-L33). It uses command line version of the linter [markdownlint-cli2](https://github.com/DavidAnson/markdownlint-cli2) that is why the config file `.markdownlint-cli2.jsonc`.

**Comment by OnkarRuikar** [comment]
Closing for now. If you have any issues let me know.

**Comment by mzbu** [comment]
@OnkarRuikar I'm taking on this issue from the OP (@nategoethel)

## Ask
I would like to search for strings that include a bolded angle bracket and replace them with the same string with an unbolded angle bracket. For example: 

Search for: `**abc > def**`

Replace with: `**abc** > **def**`

## Issue

My current code lives in the `.markdownlint.json` file and looks like this:

![image](https://user-images.githubusercontent.com/95442268/223209548-07d58c9d-58e4-41fc-a4fd-f3057093715b.png)

However, it does not flag the "search-replace" rule:

![image](https://user-images.githubusercontent.com/95442268/223211395-d1bf852e-6cb4-4732-968b-b1d5d42753ed.png)

Yet, it will still flag the other listed rules, such as MD022:

![image](https://user-images.githubusercontent.com/95442268/223211412-db6404f2-580f-4b7a-bebc-f0a02af6782b.png)

## Additional Information
I installed the extension using `npm install markdownlint-rule-search-replace --save-dev` from [npm docs](https://www.npmjs.com/package/markdownlint-rule-search-replace)

I'm quite lost as to where to go from here.

**Comment by OnkarRuikar** [comment]
> I installed the extension using npm install markdownlint-rule-search-replace --save-dev from [npm docs](https://www.npmjs.com/package/markdownlint-rule-search-replace)

Now you simply need to create one more config file `.markdownlint-cli2.jsonc` and put following text in it:
```json
{
  "customRules": ["markdownlint-rule-search-replace"]
}
```
Make sure it looks like this:
![image](https://user-images.githubusercontent.com/87750369/223303630-5b7cbb39-1935-44d4-8783-253606abfba6.png)

**Note:** Instead of maintaining two config files, it is better to move entire configuration to `.markdownlint-cli2.jsonc` and delete `.markdownlint.json`:
```json
// .markdownlint-cli2.jsonc
{
  "config": {
    "default": true,
    "MD013": false,
    ...
    "search-replace": {
      "rules": [
        {
          "name": "bold angle brackets",
          "message": "Do not bold angle brackets",
          "searchPattern": "/(\\*\\*)([A-Z]+)(\\s)(\\>)(\\s)([A-Z]+)(\\*\\*)/gi",
          "replace": "**$2** > **$6**",
          "skipCode": true
        }
      ]
    }
  },
  "customRules": ["markdownlint-rule-search-replace"]
}
```

***
## Explanation

> Rules can be enabled, disabled, and customized by creating a JSON file named `.markdownlint.jsonc`/`.markdownlint.json` ... in any directory of a project. Additionally, options (which include rules and things like markdown-it plugins and other settings) can be configured by creating a JSON file named `.markdownlint-cli2.jsonc` [[ref](https://marketplace.visualstudio.com/items?itemName=DavidAnson.vscode-markdownlint#markdownlint.config)]

In short, the VSCode markdownlint extension has various ways to provide configurations and supports multiple configuration files. We have to pick the right one. :(

Unfortunately, even though you can put custom rule config in it, the default `.markdownlint.json` file doesn't have an option to provide custom rule name/path. On other hand, the `.markdownlint-cli2.jsonc` does have option for providing custom rule name.
You can look at [MDN Web Docs'](https://github.com/mdn/content/blob/main/.markdownlint-cli2.jsonc)  setup. There we use it in VSCode, command line, and GitHub workflows.

**Comment by mzbu** [comment]
@OnkarRuikar Thank you! This fixed it!
